File: frontend-angular/libs/mocks/src/lib/mocks/file.ts
/* istanbul ignore file */

/**
 * Copyright (C) 2016-2020 TU Muenchen and contributors of ANEXIA Internetdienstleistungs GmbH
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import { mockUser } from '@eworkbench/mocks';
import { File } from '@eworkbench/types';

export const mockFile: File = {
  version_number: 4,
  content_type: 30,
  created_by: mockUser,
  metadata: [],
  location: '',
  deleted: false,
  last_modified_at: '2020-01-15T10:21:59.021913+01:00',
  name: 'ccitt_4.tif',
  created_at: '2019-10-08T18:50:55.547200+02:00',
  description: '',
  file_size: 69554,
  url: 'http://workbench.local:8000/api/files/fff1d25f-f41c-42e0-93fb-bfb83c3411a5/',
  container_id: null,
  title: 'TIFF file - bhagmann2 has no view access',
  mime_type: 'image/tiff',
  imported: false,
  content_type_model: 'shared_elements.file',
  directory_id: 'c82001c8-d78a-4364-a99b-40be1144b1be',
  original_filename: 'ccitt_4.tif',
  projects: ['21f5caf4-c7cc-4e65-8912-6ff1ae5a2478'],
  is_dss_file: false,
  download: 'http://workbench.local:8000/api/files/1128c316-d046-4a8a-9553-2c0e70f6927d/download/',
  envelope_id: null,
  display: 'ccitt_4.tif',
  pk: 'fff1d25f-f41c-42e0-93fb-bfb83c3411a5',
  last_modified_by: mockUser,
  is_favourite: false,
};

File: backend-django/app/eric/shared_elements/rest/serializers/file.py
#
# Copyright (C) 2016-2020 TU Muenchen and contributors of ANEXIA Internetdienstleistungs GmbH
# SPDX-License-Identifier: AGPL-3.0-or-later
#
from django.contrib.auth import get_user_model
from django.db import transaction
from django.urls import reverse
from django_userforeignkey.request import get_current_request
from rest_framework import serializers

from eric.core.rest.serializers import BaseModelWithCreatedBySerializer, BaseModelWithCreatedByAndSoftDeleteSerializer
from eric.dss.models import DSSEnvelope, DSSContainer
from eric.metadata.rest.serializers import EntityMetadataSerializerMixin, EntityMetadataSerializer
from eric.projects.rest.serializers.project import ProjectPrimaryKeyRelatedField
from eric.shared_elements.models import File, UploadedFileEntry

User = get_user_model()


class UploadedFileEntrySerializer(BaseModelWithCreatedBySerializer):
    """ REST API Serializer for Uploaded File Entries """

    class Meta:
        model = UploadedFileEntry
        fields = ('mime_type', 'original_filename', 'created_by', 'created_at', 'file_size')
        read_only_fields = ('mime_type', 'original_filename', 'file_size')


class FileSerializer(BaseModelWithCreatedByAndSoftDeleteSerializer, EntityMetadataSerializerMixin):
    """ REST API Serializer for Files """
    from eric.drives.models import Directory

    projects = ProjectPrimaryKeyRelatedField(many=True, required=False)

    download = serializers.SerializerMethodField(read_only=True)

    metadata = EntityMetadataSerializer(
        read_only=False,
        many=True,
        required=False,
    )

    def get_download(self, file):
        request = get_current_request()
        path = reverse("file-download", kwargs={'pk': file.uploaded_file_entry.pk})
        return request.build_absolute_uri(path)

    directory_id = serializers.PrimaryKeyRelatedField(
        # ToDo: We should also provide Directory.objects.viewable() here
        queryset=Directory.objects.all(),
        source='directory',
        many=False,
        required=False,
        allow_null=True
    )

    envelope_id = serializers.PrimaryKeyRelatedField(
        source='directory.drive.envelope',
        many=False,
        required=False,
        allow_null=True,
        read_only=True
    )

    container_id = serializers.PrimaryKeyRelatedField(
        source='directory.drive.envelope.container',
        many=False,
        required=False,
        allow_null=True,
        read_only=True
    )

    path = serializers.FileField(write_only=True, allow_empty_file=True)

    class Meta:
        model = File
        fields = (
            'title', 'name', 'description', 'path', 'original_filename', 'mime_type', 'projects', 'url',
            'download', 'file_size', 'directory_id', 'envelope_id', 'container_id', 'is_dss_file', 'location',
            'created_by', 'created_at', 'last_modified_by', 'last_modified_at', 'version_number',
            'metadata', 'imported', 'is_favourite'
        )
        read_only_fields = ('original_filename', 'mime_type', 'file_size', 'mime_type')

    @transaction.atomic
    def create(self, validated_data):
        metadata_list = self.pop_metadata(validated_data)
        instance = super().create(validated_data)
        self.create_metadata(metadata_list, instance)
        return instance

    @transaction.atomic
    def update(self, instance, validated_data):
        metadata_list = self.pop_metadata(validated_data)
        self.update_metadata(metadata_list, instance)
        return super().update(instance, validated_data)
